Source/Core/Core/AchievementManager.cpp

```diff
@@ -204,12 +206,82 @@ void AchievementManager::ActivateDeactivateRichPresence()
       nullptr, 0);
 }
 
+void AchievementManager::DoFrame()
+{
+  if (!m_is_game_loaded)
+    return;
+  Core::RunAsCPUThread([&] {
+    rc_runtime_do_frame(
+        &m_runtime,
+        [](const rc_runtime_event_t* runtime_event) {
+          AchievementManager::GetInstance()->AchievementEventHandler(runtime_event);
+        },
+        [](unsigned address, unsigned num_bytes, void* ud) {
+          return static_cast<AchievementManager*>(ud)->MemoryPeeker(address, num_bytes, ud);
+        },
+        this, nullptr);
+  });
+  if (!m_system)
+    return;
+  u64 current_time = m_system->GetCoreTiming().GetTicks();
+  if (current_time - m_last_ping_time > SystemTimers::GetTicksPerSecond() * 120)
+    m_queue.EmplaceItem([this] { PingRichPresence(GenerateRichPresence()); });
+}
+
+u32 AchievementManager::MemoryPeeker(u32 address, u32 num_bytes, void* ud)
```

Comment:
The return value here is a bit sketchy anyway, which stems from rcheevo's apparent assumption that a read can never fail. Personally I think this is pretty bad (how are you going to differentiate a real zero in memory from a failed read?), but I'm also not sure what to do about this given the current API.

Reply:
Until rcheevos is updated to support possible failed returns, then yes, this will return zero for a failed read, and yes, achievement developers will need to be aware of such.

Reply:
Very crude, but maybe give it a different sentinel in the meantime? Something like `0xcccccccc` is less likely to collide with anything expected than `0`.

Reply:
Something like 0xcccccccc may collide with some > xyz check easily

Reply:
Oh, yeah, fair enough. Well I guess there is no good solution here then...